node/grpc/v2_metrics.go

```diff
@@ -0,0 +1,202 @@
+	registry.MustRegister(collectors.NewGoCollector())
+
+	logger.Infof("Starting metrics server at port %d", port)
+	addr := fmt.Sprintf(":%d", port)
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.HandlerFor(
+		registry,
+		promhttp.HandlerOpts{},
+	))
+	server := &http.Server{
+		Addr:    addr,
+		Handler: mux,
+	}
+
+	grpcMetrics := grpcprom.NewServerMetrics()
+	registry.MustRegister(grpcMetrics)
+	grpcServerOption := grpc.UnaryInterceptor(
+		grpcMetrics.UnaryServerInterceptor(),
+	)
+
+	storeChunksLatency := promauto.With(registry).NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace:  namespace,
+			Name:       "store_chunks_latency_ms",
+			Help:       "The latency of a StoreChunks() RPC call.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{},
+	)
+
+	storeChunksDataSize := promauto.With(registry).NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "store_chunks_data_size_bytes",
+			Help:      "The size of the data requested to be stored by StoreChunks() RPC calls.",
+		},
+		[]string{},
+	)
+
+	getChunksLatency := promauto.With(registry).NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace:  namespace,
+			Name:       "get_chunks_latency_ms",
+			Help:       "The latency of a GetChunks() RPC call.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{},
+	)
+
+	getChunksDataSize := promauto.With(registry).NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "get_chunks_data_size_bytes",
+			Help:      "The size of the data requested to be retrieved by GetChunks() RPC calls.",
+		},
+		[]string{},
+	)
+
+	dbSize := promauto.With(registry).NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "db_size_bytes",
+			Help:      "The size of the leveldb database.",
+		},
+		[]string{},
+	)
+	isAlive := &atomic.Bool{}
+	isAlive.Store(true)
+
+	return &V2Metrics{
+		logger:              logger,
+		registry:            registry,
+		server:              server,
+		grpcServerOption:    grpcServerOption,
+		storeChunksLatency:  storeChunksLatency,
+		storeChunksDataSize: storeChunksDataSize,
+		getChunksLatency:    getChunksLatency,
+		getChunksDataSize:   getChunksDataSize,
+		dbSize:              dbSize,
+		dbSizePollPeriod:    dbSizePollPeriod,
+		dbDir:               dbDir,
+		isAlive:             isAlive,
+	}, nil
+}
+
+// Start starts the metrics server.
+func (m *V2Metrics) Start() {
+	go func() {
+		err := m.server.ListenAndServe()
+		if err != nil && !strings.Contains(err.Error(), "http: Server closed") {
+			m.logger.Errorf("metrics server error: %v", err)
+		}
+	}()
+
+	if m.dbSizePollPeriod.Nanoseconds() == 0 {
+		return
+	}
+	go func() {
+		ticker := time.NewTicker(m.dbSizePollPeriod)
+
+		for m.isAlive.Load() {
+			var size int64
+			err := filepath.Walk(m.dbDir, func(_ string, info os.FileInfo, err error) error {
```

Comment:
is this thread safe?

Reply:
It is almost certainly not thread safe (i.e. if level DB deletes a file or a directory mid-walk, then `filepath.Walk()` will return an error). My hope was that if the race condition was sufficiently rare, we could still extract meaningful metrics data.

Currently, this will log an error whenever this method is unable to fetch new data. Will an error log cause problems if it triggers every once in a while? If so, should this be downgraded to a a `logger.info()` call?

Unfortunately, levelDB doesn't expose API that tells you the size of the DB (that I know of). My reasoning was that this metric would be sufficiently valuable to justify a hacky collection method.

In theory, we could have the levelDB wrapper track the quantity of data, at the cost of some extra book keeping (every DB modification would need to update a special `size` key-value pair). This wouldn't tell us the size of the files on disk (which may vary depending on things like compaction and indexes), but would give us a very good idea of the approximate size if the DB. If I implemented such a thing, it would need to be in a stand alone PR though.

Reply:
This metric is now removed.